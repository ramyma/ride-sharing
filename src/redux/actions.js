import { bookingStatuses } from '../constants';

const action = (type, payload) => ({ type, ...payload });

export const setPathStep = pathStep => action('SET_PATH_STEP', { pathStep });

export const setStation = (currentStation, nextStation) =>
  action('SET_STATION', { currentStation, nextStation });

export const addBooking = booking =>
  action('ADD_BOOKING', { booking: { name: 'Name', ...booking } });

export const startTrip = startDate => action('START_TRIP', { startDate });

export const resetTrip = booking => action('RESET_TRIP');

export const endTrip = booking => action('END_TRIP');

export const boardBooking = (id, status) =>
  action('BOARD_BOOKING', {
    id,
    status:
      status ||
      [bookingStatuses.boarded, bookingStatuses.missed][
        Math.round(Math.random() > 0.8) //80% of boarding vs missing the trip
      ]
  });

export const dropoffBooking = id => action('DROPOFF_BOOKING', { id });
