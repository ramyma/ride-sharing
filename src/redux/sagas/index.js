import {
  all,
  call,
  fork,
  put,
  putResolve,
  takeEvery,
  take,
  delay,
  spawn,
  select
} from 'redux-saga/effects';
import {
  endTrip,
  setPathStep,
  setStation,
  boardBooking,
  dropoffBooking
} from '../actions';
import { path, stationsIdx, stations } from '../../data';
import {
  bookingWithPickupStation,
  bookingWithDropoffStation,
  currentPathStepSelector,
  tripStatusSelector
} from '../selectors';
import { tripStatuses } from '../../constants';

const TOTAL_TIME = 3 * 60 * 1000;
const STATION_WAIT_TIME = 10000;
const POINT_DELAY =
  (TOTAL_TIME - STATION_WAIT_TIME * stationsIdx.length - 2) /
  (path.length - stationsIdx.length);

function* handleSetPathStep(action) {
  const { pathStep } = action;

  if (stationsIdx.includes(pathStep)) {
    const { id } = stations.find(({ pos }) => pos === pathStep) || {};

    if (Number.isInteger(id)) {
      yield putResolve(setStation(id, id + 1));

      yield fork(handleBoarding);
      yield fork(handleDropoff);
    }
  }
}

function* handleBoarding() {
  const bookingsToBoard = yield select(bookingWithPickupStation);
  for (let i in bookingsToBoard) {
    const { id } = bookingsToBoard[i];
    yield put(boardBooking(id));
  }
}

function* handleDropoff() {
  const bookingsToDropoff = yield select(bookingWithDropoffStation);

  for (let i in bookingsToDropoff) {
    const { id } = bookingsToDropoff[i];
    yield put(dropoffBooking(id));
  }
}

function* watchSetPathStep() {
  yield takeEvery('SET_PATH_STEP', handleSetPathStep);
}

function* watchStartTrip() {
  while (true) {
    const action = yield take(['START_TRIP', 'persist/REHYDRATE']);

    const tripStatus = yield select(tripStatusSelector);

    // To resume the trip in case the state was rehydrated and the trip status is started
    if (action.type === 'START_TRIP' || tripStatus === tripStatuses.started) {
      let pathStep = yield select(currentPathStepSelector);
      while (pathStep < path.length - 1) {
        yield put(setPathStep(pathStep));
        // if station point wait for a while
        if (stationsIdx.includes(pathStep)) {
          yield delay(STATION_WAIT_TIME);
        } else {
          yield delay(POINT_DELAY);
        }
        pathStep++;
      }
      yield put(setPathStep(pathStep));

      yield put(endTrip());
    }
  }
}

export default function* rootSaga() {
  const sagas = [watchSetPathStep, watchStartTrip];

  yield all(
    sagas.map(saga =>
      spawn(function*() {
        while (true) {
          try {
            yield call(saga);
            break;
          } catch (e) {
            console.log(e);
          }
        }
      })
    )
  );
}
