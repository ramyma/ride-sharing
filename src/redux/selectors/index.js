import { stations } from '../../data';
import { bookingStatuses } from '../../constants';

export const bookingsSelector = state => {
  return state.bookings;
};

export const tripSelector = state => {
  return state.trip;
};

export const currentStationSelector = state => {
  const { trip: { currentStation } = {} } = state;

  return Number.isInteger(currentStation)
    ? stations[currentStation]
    : undefined;
};

export const currentPathStepSelector = state => {
  const { trip: { pathStep } = {} } = state;

  return pathStep;
};

export const nextStationSelector = state => {
  const { trip: { nextStation } = {} } = state;

  return Number.isInteger(nextStation) ? stations[nextStation] : undefined;
};

export const driverSelector = state => {
  const { trip: { driver } = {} } = state;
  return driver;
};

export const tripStatusSelector = state => {
  const { trip: { status } = {} } = state;
  return status;
};

export const tripStartDateSelector = state => {
  const { trip: { startDate } = {} } = state;
  return startDate;
};

// TODO: add unit test
export const bookingWithPickupStation = state => {
  const bookings = bookingsSelector(state);
  const currentStation = currentStationSelector(state);

  return bookings.filter(
    ({ pickupStation, status }) =>
      status === bookingStatuses.pending &&
      currentStation &&
      currentStation.id === pickupStation
  );
};

// TODO: add unit test
export const bookingWithDropoffStation = state => {
  const bookings = bookingsSelector(state);
  const currentStation = currentStationSelector(state);

  return bookings.filter(
    ({ dropoffStation, status }) =>
      status === bookingStatuses.boarded &&
      currentStation &&
      currentStation.id === dropoffStation
  );
};
