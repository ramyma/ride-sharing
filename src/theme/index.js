import { opacify } from 'polished';

const theme = {
  breakpoints: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  },
  defaultColorModeName: 'light',
  colors: {
    background: '#fff',
    grey: '#585858',
    inverseBackground: '#1c1c1c',
    shadow: opacify(0.19, '#000'),
    textDark: '#353535',
    text: '#353535',
    primary: '#ee424a',
    success: '#4CAF50',
    danger: '#f44336',
    warning: '#ffa000',
    modes: {
      dark: {
        background: '#1c1c1c',
        grey: '#9e9e9e',
        inverseBackground: '#fff',
        shadow: opacify(0.19, '#fff'),
        text: '#fff'
      }
    }
  },
  fontSizes: [0, 12, 14, 16, 20, 24, 32, 48, 64, 72],
  sizes: [0, 2, 4, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32],
  shadows: {
    grey: 'rgba(0, 0, 0, 0.19) 0 0 10px 0px',
    greyInset: `inset rgba(0, 0, 0, 0.19) 0 0 10px 0px`,
    pink: '#e42974 0 0 3px',
    text: '0 0 rgba(0, 0, 0, 0.19)'
  },
  space: [0, 4, 8, 16, 24, 48, 96, 144, 192, 240]
};

export default theme;
