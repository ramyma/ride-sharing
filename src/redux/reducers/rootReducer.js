import { combineReducers } from 'redux';
import { default as bookingsReducer } from './bookings';
import { default as tripReducer } from './trip';

const rootReducer = combineReducers({
  bookings: bookingsReducer,
  trip: tripReducer
});

export default rootReducer;
