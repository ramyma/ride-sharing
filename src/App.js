import React from 'react';
import { Global } from '@emotion/core';
import { css, useTheme } from '@xstyled/emotion';
import TripInfo from './components/TripInfo';
import Header from './components/Header';
import Map from './components/Map';
import Bookings from './components/bookings/Bookings';
import { useSelector } from 'react-redux';
import { bookingsSelector, tripStatusSelector } from './redux/selectors';
import Flex from './components/primitives/Flex';
import Charts from './components/charts/Charts';
import { tripStatuses } from './constants';

function App() {
  const bookings = useSelector(bookingsSelector);
  const tripStatus = useSelector(tripStatusSelector);

  const {
    colors: { background, inverseBackground },
  } = useTheme();

  return (
    <>
      <Global
        styles={css`
          body,
          html {
            margin: 0 auto;
            height: 100%;
            width: 100%;
          }

          body {
            background: ${background};
            color: ${inverseBackground};
            margin: 0;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
              'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
              'Helvetica Neue', sans-serif;
            -webkit-font-smoothing: antialiased;
            -moz-osx-font-smoothing: grayscale;
          }

          select {
            border-color: ${inverseBackground};
            font-size: 16px;
          }

          h1,
          h2,
          h3,
          h4,
          h5,
          p {
            margin: 0;
          }

          #root {
            height: 100%;
            width: 100%;
          }
        `}
      />
      <Header />
      <Map />

      <Flex dir="col" grid pb={6}>
        <TripInfo mt={5} />
        <Bookings mt={5} mb={5} bookings={bookings} />
        {tripStatus === tripStatuses.ended && <Charts />}
      </Flex>
    </>
  );
}

export default App;
