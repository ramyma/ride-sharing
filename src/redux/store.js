import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web
import rootReducer from './reducers/rootReducer';
import rootSaga from './sagas';

const persistConfig = {
  key: 'root',
  storage
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const sagaMiddleware = createSagaMiddleware();

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true })
    : compose;

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware))
);

export const persistor = persistStore(store);

sagaMiddleware.run(rootSaga);
