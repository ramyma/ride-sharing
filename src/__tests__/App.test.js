import React from 'react';
import renderer from 'react-test-renderer';
import { createSerializer } from 'jest-emotion';
import * as emotion from '@xstyled/emotion';
import { ColorModeProvider, ThemeProvider } from '@xstyled/emotion';
import App from '../App';
import theme from '../theme';
import { Provider } from 'react-redux';
import rootReducer from '../redux/reducers/rootReducer';
import { createStore } from 'redux';

expect.addSnapshotSerializer(createSerializer(emotion));
jest.mock('../components/Map', () => {
  return '';
});

afterAll(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
  jest.resetAllMocks();
});

const testDate = new Date(1482363367071);
jest.spyOn(global, 'Date').mockImplementation(() => testDate);
Date.now = jest.fn(() => 1482363367071);

jest.mock('../data/users');

it('App renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={createStore(rootReducer)}>
        <ThemeProvider theme={theme}>
          <ColorModeProvider>
            <App />
          </ColorModeProvider>
        </ThemeProvider>
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
