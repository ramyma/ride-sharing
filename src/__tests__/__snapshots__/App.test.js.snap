// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
Array [
  .emotion-1 {
  background-color: var(--xstyled-colors-background,#fff);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 48px;
  margin: 0 auto;
  padding: 0 24px;
  padding-top: 8px;
  padding-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

@media (max-width:1199.98px) {
  .emotion-1 {
    max-width: 100%;
  }
}

@media (min-width:1200px) {
  .emotion-1 {
    max-width: 1200px;
  }
}

.emotion-0 {
  background-color: var(--xstyled-colors-inverseBackground,#1c1c1c);
  border-radius: 50%;
  height: 40px;
  width: 40px;
  cursor: pointer;
}

.emotion-0:disabled {
  cursor: not-allowed;
}

<header
    className="emotion-1"
    height={48}
  >
    <img
      alt="logo"
      height="60%"
      src="logo.png"
    />
    <div
      className="emotion-0"
      height={40}
      onClick={[Function]}
      width={40}
    />
  </header>,
  < />,
  .emotion-33 {
  background-color: var(--xstyled-colors-background,#fff);
  margin: 0 auto;
  padding: 0 24px;
  padding-bottom: 96px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (max-width:1199.98px) {
  .emotion-33 {
    max-width: 100%;
  }
}

@media (min-width:1200px) {
  .emotion-33 {
    max-width: 1200px;
  }
}

.emotion-26 {
  background-color: var(--xstyled-colors-background,#fff);
  margin-top: 48px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-4 {
  background-color: var(--xstyled-colors-background,#fff);
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  margin-bottom: 24px;
  padding-top: 8px;
  padding-bottom: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-2 {
  background-color: var(--xstyled-colors-background,#fff);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  color: var(--xstyled-colors-text,#353535);
  font-size: 24px;
}

.emotion-1 {
  margin-top: 4px;
  color: var(--xstyled-colors-grey,#585858);
  font-size: 16px;
}

.emotion-3 {
  background-color: var(--xstyled-colors-primary,#ee424a);
  border-color: white;
  border: 1px solid;
  border-radius: 4px;
  color: white;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  min-width: 100px;
  padding: 8px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-weight: bold;
  -webkit-transition: 280ms all ease;
  transition: 280ms all ease;
}

.emotion-3:disabled {
  cursor: not-allowed;
  opacity: 0.4;
}

.emotion-25 {
  background-color: var(--xstyled-colors-background,#fff);
  grid-column-gap: 48px;
  grid-row-gap: 24px;
  grid-template-columns: minMax(280px,1fr);
  margin-top: 24px;
  display: grid;
}

@media (min-width:768px) {
  .emotion-25 {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

.emotion-9 {
  background-color: var(--xstyled-colors-background,#fff);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-5 {
  background-color: var(--xstyled-colors-background,#fff);
  border-radius: 50%;
  width: 45px;
  height: 45px;
  margin-right: 8px;
}

.emotion-6 {
  color: var(--xstyled-colors-text,#353535);
  font-size: 16px;
  font-weight: 600;
  text-transform: capitalize;
}

.emotion-7 {
  color: var(--xstyled-colors-text,#353535);
  font-size: 14px;
  text-transform: capitalize;
}

.emotion-10 {
  background: linear-gradient( 45deg,#691b9c,#e60e64);
  background-color: var(--xstyled-colors-background,#fff);
  border-radius: 5px;
  width: 4px;
  height: 37px;
  margin-right: 8px;
}

.emotion-13 {
  background-color: var(--xstyled-colors-background,#fff);
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 98%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-21 {
  background-color: var(--xstyled-colors-background,#fff);
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-15 {
  background-color: var(--xstyled-colors-grey,#585858);
  border-radius: 4px;
  height: 12px;
  width: 10px;
  margin-right: 8px;
}

@media (min-width:768px) {
  .emotion-15 {
    height: 18px;
  }
}

.emotion-16 {
  color: var(--xstyled-colors-text,#353535);
  font-size: 14px;
}

.emotion-22 {
  color: var(--xstyled-colors-text,#353535);
  font-size: 16px;
  font-weight: 600;
}

.emotion-32 {
  background-color: var(--xstyled-colors-background,#fff);
  margin-top: 48px;
  margin-bottom: 48px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-29 {
  background-color: var(--xstyled-colors-background,#fff);
  border: 1px solid;
  border-radius: 4px;
  border-color: var(--xstyled-colors-inverseBackground,#1c1c1c);
  color: var(--xstyled-colors-inverseBackground,#1c1c1c);
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  min-width: 100px;
  padding: 8px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  font-weight: bold;
  -webkit-transition: 280ms all ease;
  transition: 280ms all ease;
}

.emotion-29:disabled {
  cursor: not-allowed;
  opacity: 0.4;
}

.emotion-29:hover {
  background-color: var(--xstyled-colors-inverseBackground,#1c1c1c);
  color: var(--xstyled-colors-background,#fff);
}

<div
    className="emotion-33"
    dir="col"
  >
    <section
      className="emotion-26"
      dir="col"
    >
      <div
        className="emotion-4"
      >
        <div
          className="emotion-2"
          dir="col"
        >
          <h1
            className="emotion-0"
          >
            Trip Information
          </h1>
          <h2
            className="emotion-1"
          >
            22 / 12 / 2016, 01:36 AM
          </h2>
        </div>
        <button
          className="emotion-3"
          color="white"
          disabled={true}
          height="fit-content"
          onClick={[Function]}
        >
          Start trip
        </button>
      </div>
      <div
        className="emotion-25"
      >
        <div
          className="emotion-9"
        >
          <img
            className="emotion-5"
            height={45}
            src="img"
            width={45}
          />
          <div
            className="emotion-2"
            dir="col"
          >
            <p
              className="emotion-6"
            >
              name
            </p>
            <p
              className="emotion-7"
            >
              Car Model
               - 
              Plate
            </p>
          </div>
        </div>
        <div
          className="emotion-9"
        >
          <div
            className="emotion-10"
            height="37px"
            width={2}
          />
          <div
            className="emotion-13"
            dir="col"
            height="98%"
          >
            <p
              className="emotion-6"
            >
              Station 1
            </p>
            <p
              className="emotion-6"
            >
              Station 10
            </p>
          </div>
        </div>
        <div
          className="emotion-21"
          dir="col"
        >
          <div
            className="emotion-9"
          >
            <div
              className="emotion-15"
              height={
                Object {
                  "md": 8,
                  "xs": 5,
                }
              }
              width={4}
            />
            <p
              className="emotion-16"
            >
              Trip Distance: 
              47
               KM
            </p>
          </div>
          <div
            className="emotion-9"
          >
            <div
              className="emotion-15"
              height={
                Object {
                  "md": 8,
                  "xs": 5,
                }
              }
              width={4}
            />
            <p
              className="emotion-16"
            >
              Trip Base Fare: 
              45
               
              EGP
            </p>
          </div>
        </div>
        <div
          className="emotion-2"
          dir="col"
        >
          <p
            className="emotion-22"
          >
            Current Station: 
          </p>
          <p
            className="emotion-16"
          >
            Station 1
          </p>
        </div>
      </div>
    </section>
    <section
      className="emotion-32"
      dir="col"
    >
      <div
        className="emotion-4"
      >
        <div
          className="emotion-9"
        >
          <h1
            className="emotion-0"
          >
            Bookings
          </h1>
        </div>
        <button
          className="emotion-29"
          color="inverseBackground"
          disabled={false}
          height="fit-content"
          onClick={[Function]}
        >
          Add Booking
        </button>
      </div>
      <p
        className="emotion-16"
      >
        No bookings added yet. Press 
        <b>
          Add Booking
        </b>
         to get started.
      </p>
    </section>
  </div>,
]
`;
